### PR TITLE
Stepper update methods without BoundParameters

### PR DESCRIPTION
**Pull request by FabianKlimpel**
Currently the (C)KF needs to build multiple `BoundParameters` objects in order to update the parameters and the covariance matrix in the stepper state. This PR allows implements an explicit update method for this case. The old method becomes therewith deprecated. The corresponding methods in the `MultiTrajectory` become functions that deliver the parameters in global representation in order to be passed to the stepper state.
Alongside a reinitialisation of the Jacobian in the `AtlasStepper` is removed in order to synchronise the behaviour of the `update()` method with the other steppers.

### CODE REVIEW

File: Core/include/Acts/EventData/MultiTrajectory.ipp [hunk 1]
```diff
@@ -93,55 +93,48 @@ inline auto TrackStateProxy<SL, N, M, ReadOnly>::predictedCovariance() const
...
 template <typename SL, size_t N, size_t M, bool ReadOnly>
 inline auto TrackStateProxy<SL, N, M, ReadOnly>::filtered() const
     -> Parameters {
   assert(data().ifiltered != IndexData::kInvalid);
   return Parameters(m_traj->m_params.col(data().ifiltered).data());
 }
 
+template <typename SL, size_t N, size_t M, bool ReadOnly>
+inline auto TrackStateProxy<SL, N, M, ReadOnly>::filtered(
+    const Acts::GeometryContext& gctx) const -> FreeVector {
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/MultiTrajectory.ipp:105]: The only thing that differentiates between this method and the other `filtered()` method in the overload resolution is the argument. Can we change the name of these free parameter getters to something like `filteredGlobal`, or `filteredFree` et al?

File: Core/include/Acts/EventData/detail/coordinate_transformations.hpp [hunk 2]
```diff
@@ -126,9 +126,29 @@ struct coordinate_transformation {
...
+      const Surface& surface) {
+    FreeVector result;
+    result.template segment<3>(eFreePos0) =
+        parameters2globalPosition(gtcx, parameters, surface);
+    result[eFreeTime] = parameters[eBoundTime];
+    result.template segment<3>(eFreeDir0) =
+        parameters2globalMomentum(parameters).normalized();
+    result[eFreeQOverP] = parameters[eBoundQOverP];
+    return result;
+  }
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/detail/coordinate_transformations.hpp:148]: This is extracted from `BoundParameters`, yes?

**Comment by paulgessinger** [comment]
I think this is a good idea. I'll let @msmk0 do the actual approval, though.

File: Core/include/Acts/EventData/MultiTrajectory.ipp (hunk 1 above)
Comment (FabianKlimpel) [reply Core/include/Acts/EventData/MultiTrajectory.ipp:105]: Of course, I will change it

File: Core/include/Acts/EventData/detail/coordinate_transformations.hpp (hunk 2 above)
Comment (FabianKlimpel) [reply Core/include/Acts/EventData/detail/coordinate_transformations.hpp:148]: What do you mean with `extracted`? The methods called in here are defined in this file.

Comment (paulgessinger) [reply Core/include/Acts/EventData/detail/coordinate_transformations.hpp:148]: The logic existed before somewhere, right?. I believe `parameters2globalPosition` and `parameters2globalMomentum` were called from the `BoundParameters` constructor, is that correct?

Comment (FabianKlimpel) [reply Core/include/Acts/EventData/detail/coordinate_transformations.hpp:148]: Ah yes, there you convert it such that you have the local parameters in the `ParameterSet` and the global ones for easy access as members in the `SingleTrackParameters`

File: Core/include/Acts/EventData/MultiTrajectory.ipp (hunk 1 above)
Comment (msmk0) [reply Core/include/Acts/EventData/MultiTrajectory.ipp:105]: I agree with @paulgessinger. This should be named `filteredFree` or `freeFiltered`. My personal preference is on the latter as it groups the different accessors nicer together, i.e. `free{Filtered,Smoothed,...}`.

Comment (FabianKlimpel) [reply Core/include/Acts/EventData/MultiTrajectory.ipp:105]: Changed to `freeXy`

File: Core/include/Acts/EventData/detail/coordinate_transformations.hpp [hunk 3]
```diff
@@ -126,9 +126,29 @@ struct coordinate_transformation {
...
+  /// @return FreeVector representation of @p parameters
+  static FreeVector boundParameters2freeParameters(
+      const GeometryContext& gtcx, const BoundVector& parameters,
+      const Surface& surface) {
+    FreeVector result;
+    result.template segment<3>(eFreePos0) =
+        parameters2globalPosition(gtcx, parameters, surface);
+    result[eFreeTime] = parameters[eBoundTime];
+    result.template segment<3>(eFreeDir0) =
+        parameters2globalMomentum(parameters).normalized();
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Core/include/Acts/EventData/detail/coordinate_transformations.hpp]: done

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp [hunk 4]
```diff
@@ -316,9 +316,14 @@ BOOST_AUTO_TEST_CASE(eigen_stepper_test) {
   // Update in context of a surface
   BoundParameters bpTarget(tgContext, 2. * cov, 2. * pos, 2. * mom,
                            -1. * charge, 2. * time, targetSurface);
-  es.update(esState, bpTarget);
+  Vector3D dir = bpTarget.momentum().normalized();
+  FreeVector freeParams;
+  freeParams << bpTarget.position()[0], bpTarget.position()[1],
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/UnitTests/Core/Propagator/StepperTests.cpp]: done

File: Tests/UnitTests/Core/Propagator/StraightLineStepperTests.cpp [hunk 5]
```diff
@@ -233,9 +233,14 @@ BOOST_AUTO_TEST_CASE(straight_line_stepper_test) {
   // Update in context of a surface
   BoundParameters bpTarget(tgContext, 2. * cov, 2. * pos, 2. * mom,
                            -1. * charge, 2. * time, targetSurface);
-  sls.update(slsState, bpTarget);
+  Vector3D dir = bpTarget.momentum().normalized();
+  FreeVector freeParams;
+  freeParams << bpTarget.position()[0], bpTarget.position()[1],
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Tests/UnitTests/Core/Propagator/StraightLineStepperTests.cpp]: done

**Comment by FabianKlimpel** [comment]
> This looks good and should go in. I only have a few minor changes.
> 
> Regarding the additional free accessor methods for the `MultiTrajectory`: all the information is already available in the `MultiTrajectory` and the accessor are just there for convenience. Instead of adding them, we could just use the transformation function directly in the places where the free parameters are needed. This reduces coupling between the different parts and should be simpler to maintain. However, I am Ok with keeping this in as-is for now.
> 
> P.S. My apologies for getting this in a bit late.

I moved the functions out of the `MultiTrajectory` and built separate functions. Reducing the functionality of this class looks like the way to go.

**Pull request merged**